content/blog/2020-12-18-brilc.md

```diff
@@ -0,0 +1,248 @@
++++
+title="Struct Extension for Compiled Bril Programs"
+[[extra.authors]]
+name = "Mark Moeller"
+[extra]
+latex = true
+bio = """
+  Mark Moeller is a first-year graduate student in Computer Science at Cornell
+  interested in compilers and formal foundations of programming languages.
+"""
++++
+
+### Motivation: Garbage Collection for Compilers
+
+Building garbage collectors for compilers (as opposed to interpreters) is
+particularly tricky, since almost everything about garbage collection algorithms
+is informed by dynamic information. Assuming we are not in a virtual machine
+environment (such as Java, which can relegate details of garbage collection to
+the JVM), we must "bake in" statically a host of things to be checked or
+performed dynamically. This involves, at least, the following decisions:
+* What type of collection? While the two main choices here are tracing and
+  reference counting, we saw in the [Unified Theory of Garbage
+  Collection][unified], complete answers lie on a spectrum. This choice is somewhat
+  orthogonal to the compiler/interpreter issue.
+* When to collect? Certainly this decision will ideally be informed by dynamic
+  information about the heap state to minimize unnecessary pauses or overhead.
+  Without hardware or VM support, we may be limited in the choices of "hooks"
+  into the code to even choose from. For example, to be a drop-in replacement for
+  manual memory managment in C, one (naively at least) has only malloc calls to
+  work with as possible collection times.
+* How much to collect? Both tracing and reference counting have tunable early
+  returns that offer sacrificing memory efficiency for performance. Again,
+  whatever decisions we need to make must be evaluated by code that was emitted
+  statically.
+
+### Contributions
+
+To explore these issues, I built a new compiler from Bril to LLVM that supports the
+[memory extension][memex] (due to Zagieboylo, Doenges).
+
+Before getting into any garbage collection, I needed a little more expressive
+language for heap manipulation than just pointers to primitive arrays. For this,
+I built [a compiler][brilc] that implements a new *struct extention* for Bril, which allows
+the use of structs similar to those in C.
```

Comment:
These are super cool by themselves… if you're open to it, it would be great to merge the struct extension into the main Bril repository, and having a compiler to C there certainly couldn't hurt either.

Reply:
I am definitely open to merging the parser changes and my compiler into the bril repository. To do that, at a minimum I would want to clean up my code at least a little more (I rushed some decisions toward the end and used a global or two). It would also be best to update bril2txt to parallel the changes I made to bril2json. And then optionally add my struct feature to brili / write up a description page in the style of the existing ones?